### DISCUSSION TITLE
Improve count query performance

### DISCUSSION THREAD

**Discussion by rbushri** [ClickHouse/ClickHouse#55555] [Q&A]
Hi

I have a distributed table(ReplicatedMergeTree) the table contains ~250 billion rows and I run simple count queries with filters but the queries take around 40-200 Seconds (depending on the filter fields and the values), how can I improve the queries' performance?

Table structure
```
CREATE TABLE test_local ON CLUSTER '{cluster}' (
    ts UInt64,
    query_time UInt64,
    config_id Int32,
    domain String,
    site_id Int32,
    policy_id Int32,
    list_id Int32,
    category_id Int16,
    confidence_id Int8,
    action_id Int16,
    _ingest_time DateTime64 DEFAULT now64(3, 'UTC')
)
ENGINE = ReplicatedMergeTree
ORDER BY (ts, query_time, config_id)
TTL toDateTime(_ingest_time) + INTERVAL 30 DAY;

CREATE TABLE test ON CLUSTER '{cluster}' AS test_local
ENGINE = Distributed('{cluster}', my_db, test_local, rand());
```
The cardinality of the columns:
query_time - high 
ts - low (query_time round to hour)
config_id  - low
domain - high
site_id - low

Query for example:
```
select count() 
from test 
where (ts between 1681779600000 and 1681948800000 and query_time between 1681779600000 and 1681948800000 and config_id = 123) 
and (domain in ('abc.com.', 'abc.com') and site_id in (1234)) 
```
Query_log information:
```
query_duration_ms                    |61351           
read_rows                            |23239023006     
read_bytes                           |238478688673    
written_rows                         |0               
written_bytes                        |0               
result_rows                          |1               
result_bytes                         |256             
memory_usage                         |102576591        
```

when I tried to run the query with filters on ts and config_id it took >1 second and the count result was ~8 Billion but when I added the other filters it took 61 seconds and the count result was >20.
```
select count() 
from test 
where (ts between 1681779600000 and 1681948800000 and config_id = 123) 
```

I'm running  10 nodes of Clickhouse Database, (128 RAM, 45 COREs, disk 9 TB for each node)

Feel free to inquire if you require any further information from my side.

Thanks!!

**Comment by den-crane** [comment]
https://clickhouse.com/docs/en/sql-reference/statements/alter/projection

https://clickhouse.com/docs/en/guides/developer/cascading-materialized-views

**Reply by rbushri** [reply]
Thank you for your quick response!

In my scenario, the filtering criteria are dynamic and can contain almost all the columns. Consequently, if I were to create a Materialized View with all these columns, the MV would essentially replicate the majority of rows from the original table. This situation has me uncertain about the potential usefulness of the MV in this context.

**Reply by den-crane** [reply]
Ah, I see.

>ts UInt64,
>ORDER BY (ts, 

You have placed very cardinal `ts` column at the leading position of the orderby (primary_index). It's a foot-gun. https://kb.altinity.com/engines/mergetree-table-engine-family/pick-keys/

I suggest to use

```
partition by toYYYYMM(fromUnixTimestamp64Milli(ts))
order by (config_id, domain, ts)
```

or

```
partition by toYYYYMM(fromUnixTimestamp64Milli(ts))
primary key (intDiv(ts,86400000), config_id, domain)
order by (intDiv(ts,86400000), config_id, domain,ts)
```

**Reply by rbushri** [reply]
In my scenario, the "ts" column exhibits low cardinality with only 720 unique values (representing hourly time data over a 30-day period 24*30). On the other hand, the "query_time" column has high cardinality. My question is whether partitioning by the "ts" column could enhance query performance.
Assuming all my queries filter the data by ts, query_time, and config_id

**Reply by den-crane** [reply]
> My question is whether partitioning by the "ts" column could enhance query performance.

Yes, if you usually query only subset, 1 or 5 or 10 days of 30 days. If you query 30 days of 30 then daily partitioning can worsen performance.
